src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -25,252 +26,1010 @@
 #include "precompiled.hpp"
 #include "gc/shared/tlab_globals.hpp"
 #include "gc/shenandoah/shenandoahFreeSet.hpp"
+#include "gc/shenandoah/shenandoahFreeSet.inline.hpp"
 #include "gc/shenandoah/shenandoahHeap.inline.hpp"
 #include "gc/shenandoah/shenandoahHeapRegionSet.hpp"
 #include "gc/shenandoah/shenandoahMarkingContext.inline.hpp"
 #include "logging/logStream.hpp"
 #include "memory/resourceArea.hpp"
 #include "runtime/orderAccess.hpp"
 
-ShenandoahFreeSet::ShenandoahFreeSet(ShenandoahHeap* heap, size_t max_regions) :
-  _heap(heap),
-  _mutator_free_bitmap(max_regions, mtGC),
-  _collector_free_bitmap(max_regions, mtGC),
-  _max(max_regions)
+static const char* partition_name(ShenandoahFreeSetPartitionId t) {
+  switch (t) {
+    case NotFree: return "NotFree";
+    case Mutator: return "Mutator";
+    case Collector: return "Collector";
+    default: return "Unrecognized";
+  }
+}
+
+size_t ShenandoahSimpleBitMap::count_leading_ones(ssize_t start_idx) const {
+  assert((start_idx >= 0) && (start_idx < _num_bits), "precondition");
+  size_t array_idx = start_idx >> LogBitsPerWord;
+  uintx element_bits = _bitmap[array_idx];
```

Comment:
Same here, try to stick with what is done in bitMap.hpp (typedef uintptr_t bm_word_t, you could just as well typedef uintx bm_word_t, if you prefer).

Reply:
This is one of the "philosophical" and "tactical" differences between ShenandoahSimpleBitMap and BitMap.  ShenandoahSimpleBitMap allows searches to go in the backward direction, and uses Sentinel value -1 to indicate NotFound, and to represent "NoLimitOnBackwardSearch".  BitMap only searches in the forward direction.  Thus, it doesn't need signed values.

I could introduce a typedef to make future translation simpler if you think that would be useful.